vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -3,161 +3,180 @@
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: provisioning_bmh
+  
   - name: Get provisioned BMH names and UUIDs mapping before upgrade in KubeadmControlPlane node reuse test scenario.
-    shell: |
-        kubectl get bmh -A -o json | jq '.items[]| select (.status.provisioning.state == "provisioned")
-        | "metal3/"+.metadata.name+"="+"metal3://"+.metadata.uid' | cut -f2 -d\" | sort > /tmp/before_upgrade_mapping.txt
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
-  - name: Update maxSurge field in KubeadmControlPlane to 0 to start testing KCP scale-in feature. 
-    shell: |
-        kubectl patch -n "{{ NAMESPACE }}" kubeadmcontrolplane "{{ CLUSTER_NAME }}" --type=merge -p '{"spec": {"rolloutStrategy": {"rollingUpdate": {"maxSurge": 0}}}}'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
+    set_fact:
+      bmh_before_upgrade_kcp: "{{ bmh_before_upgrade_kcp | default([]) + [item.metadata.name + '=metal3://' + item.metadata.uid] }}"
+    with_items: "{{ provisioning_bmh.resources | filter_provisioning('provisioned') }}"
+  
   - name: Update Metal3MachineTemplate nodeReuse field to 'True'.
-    shell: |
-        kubectl get m3mt "{{ CLUSTER_NAME }}"-controlplane -n "{{ NAMESPACE }}" -ojson | jq '.spec.nodeReuse=true' | kubectl apply -f-
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
-  - name: Upgrade KubeadmControlPlane k8s version from "{{ KUBERNETES_VERSION }}" to "{{ UPGRADED_K8S_VERSION }}" to test KCP scale-in feature.
-    shell: |
-        kubectl get kubeadmcontrolplane "{{ CLUSTER_NAME }}" -n "{{ NAMESPACE }}" -o json | jq '.spec.version="{{ UPGRADED_K8S_VERSION }}"' | kubectl apply -f-
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
+    k8s:
+      api_version: infrastructure.cluster.x-k8s.io/v1alpha4
+      kind: Metal3MachineTemplate
+      name: "{{ CLUSTER_NAME }}-controlplane"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      apply: yes
+      definition:
+        spec:
+          nodeReuse: True
+  
+  - name: Update maxSurge field to 0 and upgrade KubeadmControlPlane k8s version from "{{ KUBERNETES_VERSION }}" to "{{ UPGRADED_K8S_VERSION }}" to start testing KCP scale-in feature. 
+    k8s:
+      api_version: controlplane.cluster.x-k8s.io/v1alpha3
+      kind: KubeadmControlPlane
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      apply: yes
+      definition:
+        spec:
+          rolloutStrategy:
+            rollingUpdate:
+              maxSurge: 0
+          version: "{{ UPGRADED_K8S_VERSION }}"
+  
   - pause:
       minutes: 1
```

Comment:
Last run has been hanging here forever and upgrade has proceeded well. Wondering why it should hang when you specify exact time? We can remove pause in the next run and see again.

Reply:
I think the pause was OK but it got stuck on the next step.
```
Pausing for 60 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)

TASK [v1aX_integration_test : pause] *******************************************
ok: [localhost]
```

`ok: [localhost]` should mean that it is done with the task and moving on to the next. Not sure why it is hanging there though...

Reply:
Hmm, in that case it is hanging forever while
`- name: Check if only single machine is in Deleting state and no other new machine is in Provisioning state while upgrade triggered to test the KCP scale-in feature. `

I was inside VM while last test was running and it did exactly delete 1 machine and no provisioning machine was there, but it hang, so it means there is a problem with check either it is a one time issue.

Reply:
I think this is a one time issue since it does not used to hangs there and I have just tested locally it failed in  Wait until "1" BMH is in deprovisioning state 2. same as yesterday !

Reply:
In fact, this issue is frequent although it is not always I have just increased the delay from 2 to 4 and I am testing I am not sure if this could have an impact